File: frontend/src/stores/training.ts
import { create } from 'zustand';
import { devtools, persist } from 'zustand/middleware';
import { apiClient } from '@/api/client';

export interface Skill {
  id: string;
  name: string;
  code: string;
  description?: string;
  skill_category: string;
  proficiency_levels: string[];
  minimum_required_level: number;
  is_safety_critical: boolean;
  is_quality_critical: boolean;
  requires_recertification: boolean;
  recertification_interval_days: number;
}

export interface Training {
  id: string;
  title: string;
  description?: string;
  training_type: string;
  skill_id?: string;
  instructor_id?: string;
  location?: string;
  start_date: string;
  end_date: string;
  status: string;
  capacity?: number;
  enrolled_count: number;
}

export interface TrainingRecord {
  id: string;
  user_id: string;
  user_name: string;
  training_id: string;
  training_title: string;
  status: string;
  enrolled_at: string;
  completed_at?: string;
  score?: number;
}

export interface UserSkill {
  id: string;
  user_id: string;
  skill_id: string;
  skill_name: string;
  proficiency_level: number;
  certified_at?: string;
  expires_at?: string;
  status: string;
}

interface TrainingState {
  skills: Skill[];
  trainings: Training[];
  records: TrainingRecord[];
  userSkills: UserSkill[];
  isLoading: boolean;
  error: string | null;

  fetchSkills: () => Promise<void>;
  fetchTrainings: () => Promise<void>;
  fetchRecords: () => Promise<void>;
  fetchUserSkills: (userId?: string) => Promise<void>;
  enrollInTraining: (trainingId: string | number, userId: string, notes?: string) => Promise<void>;
  registerCertification: (
    params: {
      userId: string;
      skillId: string | number;
      proficiency: number | string;
      issueDate?: string;
      expiryDate?: string;
      certificateNumber?: string;
      notes?: string;
    }
  ) => Promise<void>;
  clearError: () => void;
}

export const useTrainingStore = create<TrainingState>()(
  devtools(
    persist(
      (set) => ({
        skills: [],
        trainings: [],
        records: [],
        userSkills: [],
        isLoading: false,
        error: null,

        fetchSkills: async () => {
          set({ isLoading: true, error: null });
          try {
            const data = await apiClient.get<any>('/training/skills');
            set({ skills: data.items || [], isLoading: false });
          } catch (error: any) {
            set({ error: error.message || 'Failed to fetch skills', isLoading: false });
          }
        },

        fetchTrainings: async () => {
          set({ isLoading: true, error: null });
          try {
            const data = await apiClient.get<any>('/training/trainings');
            set({ trainings: data.items || [], isLoading: false });
          } catch (error: any) {
            set({ error: error.message || 'Failed to fetch trainings', isLoading: false });
          }
        },

        fetchRecords: async () => {
          set({ isLoading: true, error: null });
          try {
            // This might need a different endpoint or query params
            const data = await apiClient.get<any>('/training/trainings/participants');
            set({ records: data.items || [], isLoading: false });
          } catch (error: any) {
            set({ error: error.message || 'Failed to fetch records', isLoading: false });
          }
        },

        fetchUserSkills: async (userId) => {
          set({ isLoading: true, error: null });
          try {
            const url = userId ? `/training/user-skills?user_id=${userId}` : '/training/user-skills';
            const data = await apiClient.get<any>(url);
            set({ userSkills: data.items || [], isLoading: false });
          } catch (error: any) {
            set({ error: error.message || 'Failed to fetch user skills', isLoading: false });
          }
        },

        enrollInTraining: async (trainingId, userId, notes) => {
          set({ isLoading: true, error: null });
          try {
            await apiClient.post<any>(`/training/trainings/${trainingId}/participants`, {
              user_id: userId,
              notes,
            });
            // Optional: caller can refresh trainings to update enrolled_count
            set({ isLoading: false });
          } catch (error: any) {
            set({ error: error.message || 'Failed to enroll in training', isLoading: false });
            throw error;
          }
        },

        registerCertification: async ({ userId, skillId, proficiency, issueDate, expiryDate, certificateNumber, notes }) => {
          set({ isLoading: true, error: null });
          try {
            // Step 1: find or create user-skill
            let userSkillId: number | undefined;
            try {
              const list = await apiClient.get<any>(`/training/user-skills?user_id=${userId}&skill_id=${skillId}`);
              const existing = list?.items?.[0];
              if (existing?.id) {
                userSkillId = existing.id as number;
              }
            } catch {
              // listing may fail due to filters; proceed to create
            }

            if (!userSkillId) {
              const created = await apiClient.post<any>('/training/user-skills', {
                user_id: userId,
                skill_id: Number(skillId),
                proficiency_level: Number(proficiency) || 0,
                notes,
              });
              // unwrapResponse returns inner data when using APIResponse
              userSkillId = created?.id ?? created?.data?.id;
            }

            if (!userSkillId) {
              throw new Error('Unable to resolve user skill record');
            }

            // Step 2: certify the user-skill
            await apiClient.post<any>(`/training/user-skills/${userSkillId}/certify`, {
              proficiency_level: Number(proficiency) || 0,
              expiration_date: expiryDate || undefined,
              certificate_number: certificateNumber || undefined,
              notes,
            });

            // Optional: caller can refresh user skills cache if needed
            set({ isLoading: false });
          } catch (error: any) {
            set({ error: error.message || 'Failed to register certification', isLoading: false });
            throw error;
          }
        },

        clearError: () => set({ error: null }),
      }),
      {
        name: 'training-storage',
        partialize: (state) => ({
          skills: state.skills,
          trainings: state.trainings,
        }),
      }
    ),
    { name: 'TrainingStore' }
  )
);

File: sensei-rs/crates/sensei-frontend/src/stores/training.rs
//! Training reactive store.
//!
//! Mirrors the Zustand [`training.ts`](frontend/src/stores/training.ts) store.

use crate::api::client::{ApiClient, ApiError};
use leptos::prelude::*;
use serde::{Deserialize, Serialize};

/// A skill definition.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SkillDto {
    pub id: String,
    pub name: String,
    pub category: Option<String>,
    pub description: Option<String>,
    pub created_at: Option<String>,
}

/// A training course / program.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TrainingDto {
    pub id: String,
    pub title: String,
    pub description: Option<String>,
    pub skill_id: Option<String>,
    pub duration_hours: Option<f64>,
    pub training_type: Option<String>,
    pub status: Option<String>,
    pub created_at: Option<String>,
}

/// A training record (enrollment / completion).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TrainingRecordDto {
    pub id: String,
    pub user_id: String,
    pub training_id: String,
    pub status: String,
    pub score: Option<f64>,
    pub completed_at: Option<String>,
    pub notes: Option<String>,
    pub created_at: Option<String>,
}

/// A user's skill proficiency record.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UserSkillDto {
    pub id: String,
    pub user_id: String,
    pub skill_id: String,
    pub proficiency: String,
    pub certified: bool,
    pub certificate_number: Option<String>,
    pub issue_date: Option<String>,
    pub expiry_date: Option<String>,
    pub created_at: Option<String>,
}

/// Reactive store for training and skills data.
#[derive(Debug, Clone)]
pub struct TrainingStore {
    /// List of skills.
    pub skills: RwSignal<Vec<SkillDto>>,
    /// List of training courses.
    pub trainings: RwSignal<Vec<TrainingDto>>,
    /// Training records.
    pub records: RwSignal<Vec<TrainingRecordDto>>,
    /// User skill records.
    pub user_skills: RwSignal<Vec<UserSkillDto>>,
    /// Whether a fetch is in flight.
    pub is_loading: RwSignal<bool>,
    /// Last error, if any.
    pub error: RwSignal<Option<String>>,
}

impl TrainingStore {
    pub fn new() -> Self {
        Self {
            skills: RwSignal::new(Vec::new()),
            trainings: RwSignal::new(Vec::new()),
            records: RwSignal::new(Vec::new()),
            user_skills: RwSignal::new(Vec::new()),
            is_loading: RwSignal::new(false),
            error: RwSignal::new(None),
        }
    }

    /// Fetch all skills.
    pub async fn fetch_skills(&self, client: &ApiClient) {
        self.is_loading.set(true);
        self.error.set(None);
        match client.get::<Vec<SkillDto>>("/api/v1/training/skills").await {
            Ok(data) => self.skills.set(data),
            Err(e) => self.error.set(Some(e.to_string())),
        }
        self.is_loading.set(false);
    }

    /// Fetch all training courses.
    pub async fn fetch_trainings(&self, client: &ApiClient) {
        self.is_loading.set(true);
        self.error.set(None);
        match client.get::<Vec<TrainingDto>>("/api/v1/training/courses").await {
            Ok(data) => self.trainings.set(data),
            Err(e) => self.error.set(Some(e.to_string())),
        }
        self.is_loading.set(false);
    }

    /// Fetch training records.
    pub async fn fetch_records(&self, client: &ApiClient) {
        self.is_loading.set(true);
        self.error.set(None);
        match client.get::<Vec<TrainingRecordDto>>("/api/v1/training/records").await {
            Ok(data) => self.records.set(data),
            Err(e) => self.error.set(Some(e.to_string())),
        }
        self.is_loading.set(false);
    }

    /// Fetch user skills for a given user.
    pub async fn fetch_user_skills(&self, client: &ApiClient, user_id: &str) {
        self.is_loading.set(true);
        self.error.set(None);
        match client
            .get::<Vec<UserSkillDto>>(&format!("/api/v1/training/user/{}/skills", user_id))
            .await
        {
            Ok(data) => self.user_skills.set(data),
            Err(e) => self.error.set(Some(e.to_string())),
        }
        self.is_loading.set(false);
    }

    /// Enroll a user in a training course.
    pub async fn enroll_in_training(
        &self,
        client: &ApiClient,
        training_id: &str,
        user_id: &str,
        notes: Option<&str>,
    ) -> Result<TrainingRecordDto, ApiError> {
        let payload = serde_json::json!({
            "training_id": training_id,
            "user_id": user_id,
            "notes": notes,
        });
        let record: TrainingRecordDto =
            client.post("/api/v1/training/enroll", &payload).await?;
        self.records.update(|r| r.push(record.clone()));
        Ok(record)
    }

    /// Register a certification for a user (two-step: find/create user-skill + certify).
    pub async fn register_certification(
        &self,
        client: &ApiClient,
        user_id: &str,
        skill_id: &str,
        proficiency: &str,
        issue_date: Option<&str>,
        expiry_date: Option<&str>,
        certificate_number: Option<&str>,
        notes: Option<&str>,
    ) -> Result<UserSkillDto, ApiError> {
        let payload = serde_json::json!({
            "user_id": user_id,
            "skill_id": skill_id,
            "proficiency": proficiency,
            "issue_date": issue_date,
            "expiry_date": expiry_date,
            "certificate_number": certificate_number,
            "notes": notes,
        });
        let user_skill: UserSkillDto =
            client.post("/api/v1/training/certify", &payload).await?;
        self.user_skills.update(|s| s.push(user_skill.clone()));
        Ok(user_skill)
    }
}

impl Default for TrainingStore {
    fn default() -> Self {
        Self::new()
    }
}
